cocotb/run.py

```diff
@@ -0,0 +1,382 @@
+                self.work_dir = absworkdir
+
+        if python_search is None:
+            python_search = []
+
+        self.python_search = python_search
+
+        self.toplevel = toplevel
+        self.toplevel_lang = toplevel_lang
+
+        if verilog_sources is None:
+            verilog_sources = []
+
+        self.verilog_sources = self.get_abs_paths(verilog_sources)
+
+        if vhdl_sources is None:
+            vhdl_sources = []
+
+        self.vhdl_sources = self.get_abs_paths(vhdl_sources)
+
+        if includes is None:
+            includes = []
+
+        self.includes = self.get_abs_paths(includes)
+
+        if defines is None:
+            defines = []
+
+        self.defines = defines
+
+        if compile_args is None:
+            compile_args = []
+
+        if extra_args is None:
+            extra_args = []
+
+        self.compile_args = compile_args + extra_args
+
+        if simulation_args is None:
+            simulation_args = []
+
+        self.simulation_args = simulation_args + extra_args
+
+        if plus_args is None:
+            plus_args = []
+
+        self.plus_args = plus_args
+        self.force_compile = force_compile
+        self.compile_only = compile_only
+
+        for arg in kwargs:
+            setattr(self, arg, kwargs[arg])
+
+        if extra_env is not None:
+            self.env = extra_env
+        else:
+            self.env = {}
+
+        if testcase is not None:
+            self.env["TESTCASE"] = testcase
+
+        if seed is not None:
+            self.env["RANDOM_SEED"] = str(seed)
+
+        self.gui = gui
+
+    def set_env(self):
+
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        lib_dir_sep = os.pathsep + self.lib_dir + os.pathsep
+        if (
+            lib_dir_sep not in self.env["PATH"]
+        ):  # without checking will add forever casing error
+            self.env["PATH"] += lib_dir_sep
+
+        self.env["PYTHONPATH"] = os.pathsep.join(sys.path)
+
+        for path in self.python_search:
+            self.env["PYTHONPATH"] += os.pathsep + path
+
+        self.env["PYTHONHOME"] = get_config_var('prefix')
+
+        self.env["TOPLEVEL"] = self.toplevel
+        self.env["COCOTB_SIM"] = "1"
+        self.env["MODULE"] = self.module
+        self.env["VERSION"] = pkg_resources.get_distribution("cocotb").version
+
+        if not os.path.exists(self.sim_dir):
+            os.makedirs(self.sim_dir)
+
+    def build_command(self):
+        raise NotImplementedError()
+
+    def run(self):
+
+        if not os.getenv("COCOTB_RESULTS_FILE"):
+            fo = tempfile.NamedTemporaryFile()
+            results_xml_file = fo.name
+            fo.close()
+            self.env["COCOTB_RESULTS_FILE"] = results_xml_file
+        else:
+            results_xml_file = os.getenv("COCOTB_RESULTS_FILE")
+
+        cmds = self.build_command()
+        self.execute(cmds)
+
+        sys.tracebacklimit = 0  # remove not needed traceback from assert
```

Comment:
Better to use the pytest `traceback_hide` thing here, which I forget the spelling of